Cleanup of Getwork/Farm Client and move to separate lib

smurfy opened the pull request:
This is a smaller and only parts of my old PR #261

What does this PR Contain:

* PoolClient class to implement new pool protocol Implementations
* PoolManager class which uses PoolClient implementations and manages reconnects + fallback pool switches.
* Getwork/Farm class implementation using PoolClient
* Move Getwork code to separate directory/library

The idea is to keep the PR small and if the concept of the PoolClient+PoolManager is good, port the Stratum Clients to PoolClient. (already done in old PR, but needs adaption to master)

jean-m-cyr commented:
I'd love to hear a high level rationale for this restructuring. I would help me better understand what is here. How is the client manager object any different from the farm object? I don't quite follow! Are you shooting for per miner pool selection?

smurfy commented:
> I'd love to hear a high level rationale for this restructuring.

* getwork+stratum  implemented their own reconnect/fallback logic  ->  PoolManager now does it for all both
* getwork directly implemented in MinerAux instead of separate class like stratum
* MinerAux duplicates a lot of work which can be reduced by using a common interface (PoolClient)

> How is the client manager object any different from the farm object?

Farm.h Farm object, manages miners, PoolManager manages the connection to the pool (getwork and later stratum)

jean-m-cyr commented:
> Farm.h Farm object, manages miners, PoolManager manages the connection to the pool (getwork and later stratum)

Great, independent actors. So why then is it PoolManager that instantiates the farm? I do see a lot of merrit in refactoring the getwork/submit stuff.

smurfy commented:
> So why then is it PoolManager that instantiates the farm?

stratum does the farm initialization, thats why i put it there, but i think you are right, it is better to do that in MinerAux and assign the farm to the manager

jean-m-cyr commented:
Ah.. Ok! I had visions of multiple pool managers, each to their own pool, managing a subset of the available miners... never mind.

jean-m-cyr commented on ethminer/MinerAux.h:
```diff
@@ -574,9 +576,11 @@ class MinerCLI
 	{
 		_out
 			<< "Work farming mode:" << endl
+#if ETH_GETWORK
```
Should the farm and stratum parameters be made mutually exclusive?

smurfy replied:
I would prefer something like this. Most stuff already is.
Also i would like to change all the argument parsing and handling to ``cxxopts``

chfast commented:
I like the idea (was thinking about similar change for some time), but some high level comments:

1. I don't think there is a need to put everything in separated libraries. I'd rather merge all network components into a single library responsible for communication with mining pools. There is also no need for build flags for enabling / disabling stratum, getwork, etc. We should always build them.

2. I'm not a big fun of PoolManager. Can this logic be for a start in the PoolClient abstract class?

smurfy commented:
@chfast 

1.  i see your point, but i would still separate the different implementations to different sub directories. 

```
libpoolprotocols
- getwork/GetworkClient
- stratum/StratumClient1
- PoolClient
-..
```

2. I'm a fan of "separating of concerns". The PoolManager primarly manages the connections. 
while the PoolClient implement the actual protocol logic. I see this as "cleaner" but if you insist, i'm happy to move the everything to PoolClient (connection related stuff) + the hashrate reporting loop can be moved "back" to MinerAux in this case.

chfast commented:
Ok, please apply 1, and keep your PoolManager changes. I will review details later on.

chfast commented:
Where are we going with this?

smurfy commented:
I rebased this PR to latest master and adpated the output changes done to stratum to getwork too.
I also moved stratum and getwork to the libpoolprotocols as discussed.

If this PR gets merged, i will adapt stratum client to extend poolclient to work within poolmanafer with a separate PR.

smurfy commented:
> Tested. Hope you don't mind the small edit

Did i miss this change while rebasing? thanks for catching it. A lot has changed in 25 days :)

jean-m-cyr commented:
I'd say you're good to go. At least for stratum. Can't test the other protocol. This looks great, organization wise. libpoolprotocols? Quite long! Maybe just libprotocols? What other kinds of protocols do we use?

smurfy commented:
> Quite long! Maybe just libprotocols? What other kinds of protocols do we use?

i'm happy to change that with the stratum adaption
